File: octopoints-kotlin/app/src/main/java/dev/tpavani/octopoints/service/impl/DBUserService.kt
package dev.tpavani.octopoints.service.impl

import dev.tpavani.octopoints.db.DbSingleton
import dev.tpavani.octopoints.db.dao.UserDao
import dev.tpavani.octopoints.db.entity.UserEntity
import dev.tpavani.octopoints.service.DBService
import dev.tpavani.octopoints.service.model.User
import dev.tpavani.octopoints.service.utils.GetListFilter

class DBUserService : DBService<User, UserEntity>() {

    override val dao: UserDao = DbSingleton.getDb().userDao()

    override fun entityMapper(model: User): UserEntity {
        return UserEntity(model.getId(), model.getUsername(), model.getWin(), model.getLose())
    }

    override suspend fun getList(filter: GetListFilter): MutableList<User> {
        if (filter.id != null) {
            val entity = dao.getUserById(filter.id!!)
            if (entity != null) {
                return listOf(User(entity.id, entity.username, entity.win, entity.lose)).toMutableList()
            }
        } else {
            return dao.getAllUsers().map { User(it.id, it.username, it.win, it.lose) }.toMutableList()
        }
        return mutableListOf()
    }
}

File: octopoints_flutter/lib/service/impl/DBUserService.dart
import 'package:octopoints_flutter/db/DbSingleton.dart';
import 'package:octopoints_flutter/db/dao/UserDao.dart';
import 'package:octopoints_flutter/db/entity/UserEntity.dart';
import 'package:octopoints_flutter/service/DBService.dart';
import 'package:octopoints_flutter/service/model/User.dart';
import 'package:octopoints_flutter/service/utils/GetListFilter.dart';

class DBUserService extends DBService<User, UserEntity>{
  @override
  UserDao get dao => DbSingleton().db.userDao;

  @override
  UserEntity entityMapper(User model) {
    return UserEntity(model.id == 0 ? null : model.id, model.username, model.win, model.lose);
  }

  @override
  Future<List<User>> getList(GetListFilter filters) async {
    if(filters.id != null){
      UserEntity? userEntity = await (dao).getUserById(filters.id!);
      if(userEntity != null){
        return [User(userEntity.username, id: userEntity.id!, win: userEntity.win, lose: userEntity.lose)];
      }
    } else {
      return (await (dao).getAllUsers()).map((userEntity) => User(userEntity.username, id: userEntity.id!, win: userEntity.win, lose: userEntity.lose)).toList();
    }
    return [];
  }

}